src/device/esp/esp32c3.S

```diff
@@ -0,0 +1,49 @@
+// This is a very minimal bootloader for the ESP32-C3. It only initializes the
+// flash and then continues with the generic RISC-V initialization code, which
```

Comment:
I see you have figured out cache initialization part already, but just in case, you might find this useful:
https://github.com/espressif/esp32c3-direct-boot-example

Reply:
Oh wow, I wasn't aware of that! That would have been quite useful!

For previous chips (ESP8266, ESP32), we basically load everything in RAM (loaded by the ROM bootloader) and just execute from there. I got that working on the ESP32-C3, but of course that's not ideal: there is only so much RAM available and it's kind of hard to make that work well with linker scripts. So I wanted to see whether I could get execution from flash working, ideally while keeping firmware size low. Luckily I managed to do that. It's a bit hacky, but it sort of works. Linker scripts are very much not made for the instruction/data bus split (with the same underlying address space) in the ESP32 family. I might make the same change for other chips in the family.

Oh, also, just as a side note, I think there is a bug in the ESP32-C3 reference manual (page 80):

> The CPU accesses the external memory via the cache. According to the MMU (Memory Management Unit)
settings, the cache maps the CPU’s address to the external memory’s physical address. Due to this address
mapping, **the ESP32-C3 can address up to 16 MB external flash**.
> 
> Using the cache, ESP32-C3 is able to support the following address space mappings. **Note that the instruction
bus address space (8MB) and the data bus address space (8 MB) is always shared.**

Emphasis mine. Those two statements appear to be contradictory. In my experiments, it appears that the address space is indeed shared and if something is mapped in the instruction bus address space, it's also mapped in the data bus address space at the same relative location. So I don't see how it's possible to access 16MB of flash that way. This 16MB is also repeated on page 76 ("Supports up to 16 MB external flash"), which also seems wrong.